pilot/cmd/pilot-discovery/server/server.go

```diff
@@ -0,0 +1,391 @@
+}
+
+// NewServer creates a new Server instance based on the provided arguments.
+func NewServer(args PilotArgs, stop chan struct{}) (*Server, error) {
+	config := &pilotConfig{stop: stop}
+
+	// Apply the arguments to the configuration.
+	if err := aggregateConfigFn(&args, config); err != nil {
+		return nil, multierror.Prefix(err, "failed to create Pilot discovery service.")
+	}
+
+	return &Server{
+		config: config,
+	}, nil
+}
+
+// Run starts the pilot discovery service.
+func (s Server) Run() {
+	// Now start all of the services.
+	for _, runFn := range s.config.runFuncs {
+		runFn()
+	}
+}
+
+// runFunc is a no-op function that will be executed after the Pilot discovery service has been configured.
+type runFunc func()
+
+// pilotConfig contains the runtime structures for the Pilot discovery service. This structure is configured
+// by a number of pilotConfigFunc functions.
+type pilotConfig struct {
+	stop              chan struct{}
+	mesh              *meshconfig.MeshConfig
+	serviceController *aggregate.Controller
+	configController  model.ConfigStoreCache
+	mixerSAN          []string
+	runFuncs          []runFunc
+}
+
+// pilotConfigFunc is a function that configures some part of the pilotConfig based on the settings in the PilotArgs.
+type pilotConfigFunc func(in *PilotArgs, out *pilotConfig) error
+
+// initMesh creates the mesh in the pilotConfig from the input arguments.
+func initMesh(in *PilotArgs, out *pilotConfig) error {
+	// If a config file was specified, use it.
+	if in.Mesh.ConfigFile != "" {
+		mesh, err := cmd.ReadMeshConfig(in.Mesh.ConfigFile)
+		if err == nil {
+			out.mesh = mesh
+			return nil
+		}
+		glog.Warningf("failed to read mesh configuration, using default: %v", err)
+	}
+
+	// Config file either wasn't specified or failed to load - use a default mesh.
+	defaultMesh := proxy.DefaultMeshConfig()
+	mesh := &defaultMesh
+
+	// Allow some overrides for testing purposes.
+	if in.Mesh.MixerAddress != "" {
+		mesh.MixerAddress = in.Mesh.MixerAddress
+	}
+	if in.Mesh.RdsRefreshDelay != nil {
+		mesh.RdsRefreshDelay = in.Mesh.RdsRefreshDelay
+	}
+
+	glog.V(2).Infof("mesh configuration %s", spew.Sdump(mesh))
+	glog.V(2).Infof("version %s", version.Line())
+	glog.V(2).Infof("flags %s", spew.Sdump(in))
+
+	out.mesh = mesh
+	return nil
+}
+
+// initMixerSan configures the mixerSAN configuration item. The mesh must already have been configured.
+func initMixerSan(in *PilotArgs, out *pilotConfig) error {
+	if out.mesh.DefaultConfig.ControlPlaneAuthPolicy == meshconfig.AuthenticationPolicy_MUTUAL_TLS {
+		out.mixerSAN = envoy.GetMixerSAN(in.Config.ControllerOptions.DomainSuffix, in.Namespace)
+	}
+	return nil
+}
+
+type mockController struct{}
+
+func (c *mockController) AppendServiceHandler(f func(*model.Service, model.Event)) error {
+	return nil
+}
+
+func (c *mockController) AppendInstanceHandler(f func(*model.ServiceInstance, model.Event)) error {
+	return nil
+}
+
+func (c *mockController) Run(<-chan struct{}) {}
+
```

Comment:
Why are these here?

Reply:
The mock controller is used when using the `mock` registry. This was copied from another test IIRC.

Reply:
The reason I ask is because we have been putting mock functions in test files, not in main files. I am not about to nitpick on go-isms :). I leave it to likes of @ZackButcher / @kyessenov / @ijsnellf / @GregHanson for those kinds of things..

Reply:
mocks generally should be in `_test` package if they are only called from testing code. it isolates their namespace to test code only.

Reply:
Given that this is not a test package, what is your suggestion for how to allow the service to be configured with mocks?

Reply:
No shortcut I can see. You can leave it here, provided we do something later:
- make mock service registry a real memory thing
- more flexible registry configuration (through pluggable configs, not structs)